src/persist-client/src/internal/encoding.rs

```diff
@@ -907,51 +936,148 @@ impl RustType<ProtoVersionedData> for VersionedData {
     }
 }
 
-impl<T: Timestamp + Lattice + Codec64> RustType<ProtoTrace> for Trace<T> {
+impl RustType<ProtoSpineId> for SpineId {
+    fn into_proto(&self) -> ProtoSpineId {
+        ProtoSpineId {
+            lo: self.0.into_proto(),
+            hi: self.1.into_proto(),
+        }
+    }
+
+    fn from_proto(proto: ProtoSpineId) -> Result<Self, TryFromProtoError> {
+        Ok(SpineId(proto.lo.into_rust()?, proto.hi.into_rust()?))
+    }
+}
+
+impl<T: Timestamp + Codec64> ProtoMapEntry<SpineId, Arc<HollowBatch<T>>> for ProtoIdHollowBatch {
+    fn from_rust<'a>(entry: (&'a SpineId, &'a Arc<HollowBatch<T>>)) -> Self {
+        let (id, batch) = entry;
+        ProtoIdHollowBatch {
+            id: Some(id.into_proto()),
+            batch: Some(batch.into_proto()),
+        }
+    }
+
+    fn into_rust(self) -> Result<(SpineId, Arc<HollowBatch<T>>), TryFromProtoError> {
+        let id = self.id.into_rust_if_some("id")?;
+        let batch = Arc::new(self.batch.into_rust_if_some("batch")?);
+        Ok((id, batch))
+    }
+}
+
+impl<T: Timestamp + Codec64> RustType<ProtoSpineBatch> for ThinSpineBatch<T> {
+    fn into_proto(&self) -> ProtoSpineBatch {
+        ProtoSpineBatch {
+            desc: Some(self.desc.into_proto()),
+            parts: self.parts.into_proto(),
+            level: self.level.into_proto(),
+        }
+    }
+
+    fn from_proto(proto: ProtoSpineBatch) -> Result<Self, TryFromProtoError> {
+        let level = proto.level.into_rust()?;
+        let desc = proto.desc.into_rust_if_some("desc")?;
```

Comment:
optional: I think we typically prefix these `into_rust_if_some` strings with the proto name, so something like `ProtoSpineBatch::desc`, for a bit of additional context if they pop up. Otoh, if we do see one of these, then something has gone horribly wrong and it'll probably be pretty easy to track down, so I don't feel strongly